currency_rate_update/models/res_currency_rate_provider.py

```diff
@@ -0,0 +1,313 @@
+        default=fields.Date.today(),
+        required=True,
+    )
+
+    _sql_constraints = [
+        (
+            'service_company_id_uniq',
+            'UNIQUE(service, company_id)',
+            'Service can only be used in one provider per company!'
+        ),
+        (
+            'valid_interval_number',
+            'CHECK(interval_number > 0)',
+            'Scheduled update interval must be greater than zero!'
+        )
+    ]
+
+    @api.onchange('service')
+    def _onchange_service(self):
+        res = {
+            'domain': {
+                'currency_ids': self._domain_currency_ids(),
+            },
+        }
+
+        return res
+
+    @api.onchange('currency_ids')
+    def _onchange_currency_ids(self):
+        res = {
+            'domain': {
+                'currency_ids': self._domain_currency_ids(),
+            },
+        }
+
+        return res
+
+    @api.multi
+    @api.depends('service')
+    def _compute_name(self):
+        for provider in self:
+            provider.name = list(filter(
+                lambda x: x[0] == provider.service,
+                self._selection_service()
+            ))[0][1]
+
+    @api.multi
+    @api.depends('active', 'interval_type', 'interval_number')
+    def _compute_update_schedule(self):
+        for provider in self:
+            if not provider.active:
+                provider.update_schedule = _('Inactive')
+                continue
+
+            provider.update_schedule = _('%(number)s %(type)s') % {
+                'number': self.interval_number,
+                'type': list(filter(
+                    lambda x: x[0] == provider.interval_type,
+                    self._fields['interval_type'].selection
+                ))[0][1],
+            }
+
+    @api.multi
+    def _domain_currency_ids(self):
+        domain = [('active', '=', True)]
+
+        if self and self.service:
+            supported_currencies = self._get_supported_currencies(
+                self.service
+            )
+            domain = domain + [
+                ('name', 'in', supported_currencies),
+            ]
+
+        return domain
+
+    @api.multi
+    def _update(self, date_from, date_to, newest_only=False):
+        Currency = self.env['res.currency']
+        CurrencyRate = self.env['res.currency.rate']
+
+        for provider in self:
+            data = self._obtain_rates(
+                provider.service,
+                provider.company_id.currency_id.name,
+                provider.currency_ids.mapped('name'),
+                date_from,
+                date_to
+            ).items()
+
+            if not data:
+                continue
+            if newest_only:
+                data = [max(
+                    data,
+                    key=lambda x: datetime.strptime(x[0], '%Y-%m-%d')
+                )]
+
+            for content_date, rates in data:
+                timestamp = datetime.strptime(content_date, '%Y-%m-%d')
```

Comment:
use `fields.Datetime.from_string` always, but even better, don't return strings now in v12!

Reply:
Why to use full datetime?

Reply:
Sorry, I meant `fields.Date.from_string`. My point is to abstract `datetime` library. This can be beneficial for things like testing (https://twitter.com/PedroMBaeza/status/1069901067004141573)